### PR TITLE
CAMEL-19429: re-add camel-activemq component

### PR DESCRIPTION
**Pull request by jbonofre**
# Description

This PR re-adds camel-activemq component, using activemq-client-jakarta supporting JMS 2.x.

# Target

- [X] I checked that the commit is targeting the correct branch (note that Camel 3 uses `camel-3.x`, whereas Camel 4 uses the `main` branch)

# Tracking
- [X] If this is a large change, bug fix, or code improvement, I checked there is a [CAMEL-19429](https://issues.apache.org/jira/browse/CAMEL-19429) filed for the change.

# Apache Camel coding standards and style

- [X] I checked that each commit in the pull request has a meaningful subject line and body.

- [X] I formatted the code using `mvn -Pformat,fastinstall install && mvn -Psourcecheck`

### CODE REVIEW

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/package.html [hunk 1]
```diff
@@ -0,0 +1,31 @@
+<!--
```
Comment (oscerd) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/package.html:1]: Not sure we need this.

**Review by oscerd — CHANGES_REQUESTED** [review]
I think the component needs to go in Parent/pom too.

**Comment by oscerd** [comment]
Looks good overall

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+import org.springframework.jms.connection.SingleConnectionFactory;
+import org.springframework.jms.core.JmsTemplate;
+
+/**
+ * The ActiveMQ Component.
+ */
+@Component("activemq")
+public class ActiveMQComponent extends JmsComponent {
+    private final CopyOnWriteArrayList<SingleConnectionFactory> singleConnectionFactoryList = new CopyOnWriteArrayList<>();
+    private final CopyOnWriteArrayList<Service> pooledConnectionFactoryServiceList = new CopyOnWriteArrayList<>();
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java:44]:
```suggestion
    private final List<SingleConnectionFactory> singleConnectionFactoryList = new CopyOnWriteArrayList<>();
    private final List<Service> pooledConnectionFactoryServiceList = new CopyOnWriteArrayList<>();
```

It looks like we don't call any specific methods so we can use `List`

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+     * Creates an <a href="http://camel.apache.org/activemq.html">ActiveMQ Component</a> connecting to the given
+     * <a href="http://activemq.apache.org/configuring-transports.html">broker URL</a>
+     *
+     * @param  brokerURL the URL to connect to
+     * @return           the created component
+     */
+    public static ActiveMQComponent activeMQComponent(String brokerURL) {
+        ActiveMQComponent answer = new ActiveMQComponent();
+        if (answer.getConfiguration() instanceof ActiveMQConfiguration) {
+            ((ActiveMQConfiguration) answer.getConfiguration()).setBrokerURL(brokerURL);
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java:76]:
```suggestion
        if (answer.getConfiguration() instanceof ActiveMQConfiguration activeMQConfiguration) {
            activeMQConfiguration.setBrokerURL(brokerURL);
```

Comment (essobedo) [reply components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java:76]: In which case can we have another type?

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+        }
+
+        return answer;
+    }
+
+    public String getBrokerURL() {
+        if (getConfiguration() instanceof ActiveMQConfiguration) {
+            return ((ActiveMQConfiguration) getConfiguration()).getBrokerURL();
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java:84]:
```suggestion
        if (getConfiguration() instanceof ActiveMQConfiguration activeMQConfiguration) {
            return activeMQConfiguration.getBrokerURL();
```

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java (hunk 3 above)
Comment (essobedo) [reply components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java:76]: I mean this test should always pass or fail

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+    }
+
+    /**
+     * Sets the broker URL to use to connect to ActiveMQ. If none configured then localhost:61616 is used by default
+     * (however can be overridden by configuration from environment variables)
+     */
+    @Metadata(label = "common")
+    public void setBrokerURL(String brokerURL) {
+        if (getConfiguration() instanceof ActiveMQConfiguration) {
+            ((ActiveMQConfiguration) getConfiguration()).setBrokerURL(brokerURL);
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java:96]:
```suggestion
        if (getConfiguration() instanceof ActiveMQConfiguration activeMQConfiguration) {
            activeMQConfiguration.setBrokerURL(brokerURL);
```

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+
+    /**
+     * Define if all Java packages are trusted or not (for Java object JMS message types). Notice its not recommended
+     * practice to send Java serialized objects over network. Setting this to true can expose security risks, so use
+     * this with care.
+     */
+    @Metadata(defaultValue = "false", label = "advanced")
+    public void setTrustAllPackages(boolean trustAllPackages) {
+        if (getConfiguration() instanceof ActiveMQConfiguration) {
+            ((ActiveMQConfiguration) getConfiguration()).setTrustAllPackages(trustAllPackages);
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java:108]:
```suggestion
        if (getConfiguration() instanceof ActiveMQConfiguration) {
            ((ActiveMQConfiguration) getConfiguration()).setTrustAllPackages(trustAllPackages);
```
```suggestion
        if (getConfiguration() instanceof ActiveMQConfiguration activeMQConfiguration) {
            activeMQConfiguration.setTrustAllPackages(trustAllPackages);
```

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+        }
+    }
+
+    public boolean isTrustAllPackages() {
+        if (getConfiguration() instanceof ActiveMQConfiguration) {
+            return ((ActiveMQConfiguration) getConfiguration()).isTrustAllPackages();
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java:114]:
```suggestion
        if (getConfiguration() instanceof ActiveMQConfiguration activeMQConfiguration) {
            return activeMQConfiguration.isTrustAllPackages();
```

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+     * from outside of a message consuming thread, pooling will be used rather than the default with the Spring
+     * {@link JmsTemplate} which will create a new connection, session, producer for each message then close them all
+     * down again.
+     * <p/>
+     * The default value is true.
+     */
+    @Metadata(defaultValue = "true", label = "common")
+    public void setUsePooledConnection(boolean usePooledConnection) {
+        if (getConfiguration() instanceof ActiveMQConfiguration) {
+            ((ActiveMQConfiguration) getConfiguration()).setUsePooledConnection(usePooledConnection);
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java:130]:
```suggestion
        if (getConfiguration() instanceof ActiveMQConfiguration activeMQConfiguration) {
            activeMQConfiguration.setUsePooledConnection(usePooledConnection);
```

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+        }
+    }
+
+    public boolean isUsePooledConnection() {
+        if (getConfiguration() instanceof ActiveMQConfiguration) {
+            return ((ActiveMQConfiguration) getConfiguration()).isUsePooledConnection();
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java:136]:
```suggestion
        if (getConfiguration() instanceof ActiveMQConfiguration activeMQConfiguration) {
            return activeMQConfiguration.isUsePooledConnection();
```

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+     * to ActiveMQ from outside of a message consuming thread, pooling will be used rather than the default with the
+     * Spring {@link JmsTemplate} which will create a new connection, session, producer for each message then close them
+     * all down again.
+     * <p/>
+     * The default value is false and a pooled connection is used by default.
+     */
+    @Metadata(defaultValue = "false", label = "common")
+    public void setUseSingleConnection(boolean useSingleConnection) {
+        if (getConfiguration() instanceof ActiveMQConfiguration) {
+            ((ActiveMQConfiguration) getConfiguration()).setUseSingleConnection(useSingleConnection);
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java:152]:
```suggestion
        if (getConfiguration() instanceof ActiveMQConfiguration activeMQConfiguration) {
            activeMQConfiguration.setUseSingleConnection(useSingleConnection);
```

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+        }
+    }
+
+    public boolean isUseSingleConnection() {
+        if (getConfiguration() instanceof ActiveMQConfiguration) {
+            return ((ActiveMQConfiguration) getConfiguration()).isUseSingleConnection();
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java:158]:
```suggestion
        if (getConfiguration() instanceof ActiveMQConfiguration activeMQConfiguration) {
            return activeMQConfiguration.isUseSingleConnection();
```

**Comment by davsclaus** [comment]
A good start. It would also be nice with a few unit test (we dont need all the old existing tests, but a set to cover some ground).

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+        }
+    }
+
+    @Override
+    protected void doStop() throws Exception {
+        for (Service s : pooledConnectionFactoryServiceList) {
+            try {
+                s.stop();
+            } catch (Exception e) {
+                // ignore
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java:223]: I would personally rather prefer to add a logger to this class and log the errors

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+            }
+        }
+        pooledConnectionFactoryServiceList.clear();
+
+        for (SingleConnectionFactory s : singleConnectionFactoryList) {
+            try {
+                s.destroy();
+            } catch (Exception e) {
+                // ignore
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java:232]: ditto

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+        super.doStop();
+    }
+
+    /**
+     * Configuration of ActiveMQ
+     */
+    @Override
+    public void setConfiguration(JmsConfiguration configuration) {
+        if (configuration instanceof ActiveMQConfiguration) {
+            ((ActiveMQConfiguration) configuration).setActiveMQComponent(this);
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQComponent.java:246]:
```suggestion
        if (configuration instanceof ActiveMQConfiguration activeMQConfiguration) {
            activeMQConfiguration.setActiveMQComponent(this);
```

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQConfiguration.java [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+    }
+
+    @Override
+    public void setConnectionFactory(ConnectionFactory connectionFactory) {
+        ActiveMQConnectionFactory acf = null;
+
+        ConnectionFactory target = connectionFactory;
+        if (target instanceof CachingConnectionFactory) {
+            CachingConnectionFactory ccf = (CachingConnectionFactory) target;
+            target = ccf.getTargetConnectionFactory();
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQConfiguration.java:145]:
```suggestion
        if (target instanceof CachingConnectionFactory ccf) {
            target = ccf.getTargetConnectionFactory();
```

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQConfiguration.java [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+        }
+        if (target instanceof DelegatingConnectionFactory) {
+            DelegatingConnectionFactory dcf = (DelegatingConnectionFactory) target;
+            target = dcf.getTargetConnectionFactory();
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQConfiguration.java:149]:
```suggestion
        if (target instanceof DelegatingConnectionFactory dcf) {
            target = dcf.getTargetConnectionFactory();
```

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQConfiguration.java [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+        }
+        if (target instanceof ActiveMQConnectionFactory) {
+            acf = (ActiveMQConnectionFactory) target;
+        }
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/ActiveMQConfiguration.java:153]: Would it make sense to use if / else if blocks instead in this case?

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/OriginalDestinationPropagateStrategy.java [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+        if (exchange.getIn() instanceof JmsMessage) {
+            JmsMessage msg = exchange.getIn(JmsMessage.class);
+            Message jms = msg.getJmsMessage();
+            if (jms != null && jms instanceof ActiveMQMessage && message instanceof ActiveMQMessage) {
+                ActiveMQMessage amq = (ActiveMQMessage) jms;
+                if (amq.getOriginalDestination() == null) {
+                    ActiveMQDestination from = amq.getDestination();
+                    if (from != null) {
+                        LOG.trace("Setting OriginalDestination: {} on {}", from, message);
+                        ((ActiveMQMessage) message).setOriginalDestination(from);
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/OriginalDestinationPropagateStrategy.java:49]:
```suggestion
        if (exchange.getIn() instanceof JmsMessage msg) {
            Message jms = msg.getJmsMessage();
            if (jms != null && jms instanceof ActiveMQMessage amq && message instanceof ActiveMQMessage activeMQMessage) {
                if (amq.getOriginalDestination() == null) {
                    ActiveMQDestination from = amq.getDestination();
                    if (from != null) {
                        LOG.trace("Setting OriginalDestination: {} on {}", from, message);
                        activeMQMessage.setOriginalDestination(from);
```

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/converter/ActiveMQMessageConverter.java [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+                return "Processor of MessageListener: " + listener;
+            }
+        };
+    }
+
+    private static ActiveMQMessage createActiveMQMessage(Exchange exchange) throws JMSException {
+        Object body = exchange.getIn().getBody();
+        if (body instanceof String) {
+            ActiveMQTextMessage answer = new ActiveMQTextMessage();
+            answer.setText((String) body);
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/converter/ActiveMQMessageConverter.java:76]:
```suggestion
        if (body instanceof String bodyAsString) {
            ActiveMQTextMessage answer = new ActiveMQTextMessage();
            answer.setText(bodyAsString);
```

File: components/camel-activemq/src/main/java/org/apache/camel/component/activemq/converter/ActiveMQMessageConverter.java [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+            return answer;
+        } else if (body instanceof Serializable) {
+            ActiveMQObjectMessage answer = new ActiveMQObjectMessage();
+            answer.setObject((Serializable) body);
```
Comment (essobedo) [inline components/camel-activemq/src/main/java/org/apache/camel/component/activemq/converter/ActiveMQMessageConverter.java:80]:
```suggestion
        } else if (body instanceof Serializable bodyAsObject) {
            ActiveMQObjectMessage answer = new ActiveMQObjectMessage();
            answer.setObject(bodyAsObject);
```

File: parent/pom.xml [hunk 21]
```diff
@@ -50,6 +50,7 @@
         <checkstyle.failOnViolation>false</checkstyle.failOnViolation>
         <!-- dependency versions -->
         <abdera-version>1.1.3</abdera-version>
+        <activemq-version>5.18.1</activemq-version>
```
Comment (essobedo) [inline parent/pom.xml:53]: I'm wondering if we need to add it also to `camel-dependencies/pom.xml`

**Review by essobedo — COMMENTED** [review]
Awesome! Thx for bringing it back! Only a few nitpick suggestions. Unless, I missed them but it looks like there are no tests so far but you added a test resource, is-it normal?

**Comment by davsclaus** [comment]
@essobedo its the old existing code copied over as-is, eg we know it works, and we need to backport changes so dont need to do java17 code style etc

**Comment by essobedo** [comment]
> @essobedo its the old existing code copied over as-is, eg we know it works, and we need to backport changes so dont need to do java17 code style etc

That is sad that we cannot improve a very old code hardly readable with `instance of` and `cast` all over the place

**Comment by jbonofre** [comment]
My intention was to restart from existing component, adapting to use the new ActiveMQ client (jakarta/JMS).

I can also do code style improvement if we agree.

About the tests, yes, I'm working on it (without adding ActiveMQ infra test as it's not required here, just an inner broker service), that's why the PR is still draft, I will move to "ready to review" when I will push the tests.

**Comment by orpiske** [comment]
> > @essobedo its the old existing code copied over as-is, eg we know it works, and we need to backport changes so dont need to do java17 code style etc
> 
> That is sad that we cannot improve a very old code hardly readable with `instance of` and `cast` all over the place

We have quite a lot of old cold in great need of modernization, indeed. 

But ... I think this is just temporary: my understanding is that the community still expect some level of fixes back-porting from 4 to 3.21. I remember having that discussion on a PR sometime ago.

**Comment by davsclaus** [comment]
Can we prioritize this as it would be good to build RC1 next week

**Comment by orpiske** [comment]
@jbonofre @davsclaus is there anything I can do to help? I can re-prioritize my work stream to help with this, if needed.

**Comment by davsclaus** [comment]
okay, we basically just need a few basic unit tests, and to copy over the doc as well (src/main/docs)

**Comment by davsclaus** [comment]
@orpiske you are welcome to help with finishing this PR and work so we can get activemq 5.x support in v4.
We need this done this week so we can make it for the RC release

**Comment by orpiske** [comment]
> @orpiske you are welcome to help with finishing this PR and work so we can get activemq 5.x support in v4. We need this done this week so we can make it for the RC release

I will take over this one, then. Thanks!

**Comment by davsclaus** [comment]
closing this PR as this has been added back

**Pull request closed without merging**